web-ui/view/build.ml

```diff
@@ -174,3 +184,169 @@ let ref_row ~ref_title ~short_hash ~last_updated ~status ~ref_uri ~message =
+                const right_matter = document.createElement("div");  // ran_for, right_arrow
+                const step_ran_for = document.createElement("div"); // ran_for
+                const right_arrow_elt = document.createElement("div");
+
+                step_title.textContent = step["variant"];
+                step_title.setAttribute("class", "text-gray-900 text-sm font-medium");
+
+                step_created_at.textContent = "Created at: " + step["created_at"];
+                hyphen.textContent = "-";
+                step_queued_for.textContent = step["queued_for"] + " in queue";
+
+                step_timestamps_durations.appendChild(step_created_at);
+                step_timestamps_durations.appendChild(hyphen);
+                step_timestamps_durations.appendChild(step_queued_for);
+                step_timestamps_durations.setAttribute("class", "flex text-sm space-x-2");
+
+                step_info.appendChild(step_title);
+                step_info.appendChild(step_timestamps_durations);
+                step_info.setAttribute("class", "flex flex-col");
+
+                step_status.innerHTML = icon_from(step["status"]);
+                step_status.setAttribute("id", step["variant"] + "-status");
+
+                main.appendChild(step_status);
+                main.appendChild(step_info);
+                main.setAttribute("class", "flex items-center space-x-3");
+
+                step_ran_for.textContent = "Ran for " + step["ran_for"];
+
+                right_arrow_elt.innerHTML = right_arrow;
+                right_matter.appendChild(step_ran_for);
+                right_matter.appendChild(right_arrow_elt);
+                right_matter.setAttribute("class", "flex text-sm font-normal text-gray-500 space-x-8 items-center");
+
+                row.appendChild(main);
+                row.appendChild(right_matter);
+                return row
+              }
+
+              const steps_table = document.getElementById("table-container");
+              steps_table.replaceChildren(); // clears the table
+
+              // TODO: Clearing the table out and appending each step row
+              // is the simplest thing I could do to get started. It seems
+              // like it should result in jankiness from the table being completely redrawn.
+              // I'm not seeing anything like that though so I am leaving it like this.
+              // I was going to do something clever like checking if a step has already
+              // been rendered and if so, modifying its data in place.
+              for (step of data["steps"]) {
+                steps_table.appendChild(build_step_row(step, data["step_route_prefix"]))
+              }
+
+              if (
+                data["status"].startsWith("passed") ||
+                data["status"].startsWith("failed")
+              ) {
+                build_ran_for.innerHTML = "Ran for " + data["ran_for"];
+                if (data["can_cancel"]) {
```

Comment:
Yes - is this not the current behaviour? This piece of code is allowing the view to be told by the build if it can be cancelled or rebuilt and it is using this data to render the cancel button or the rebuild button(s).

Reply:
I'm not sure about the meaning if we said we could cancel once the build is finished (passed, failed). We should be able to cancel the build when it's active, no?

Reply:
Yes we should. This should be the behaviour at the moment. Is it not?

Reply:
Ah ok I think I understand why I am confused. 

Facts:
- At the moment, you can cancel a _build_ when it is running
- At the moment, you cannot cancel a _step_ when it is running (you can only rebuild it). See also #554 


The code here is about cancelling a build. If I understand correctly, you are asking about why we can't cancel a step?

Reply:
It's a misunderstanding on my side, sorry! Closing it :+1: